packages/js/components/src/tree-select-control/index.js

```diff
@@ -0,0 +1,497 @@
+ *
+ * @param {Object}                     props                              Component props.
+ * @param {string}                     [props.id]                         Component id
+ * @param {string}                     [props.label]                      Label for the component
+ * @param {string | false}             [props.selectAllLabel]             Label for the Select All root element. False for disable.
+ * @param {string}                     [props.help]                       Help text under the select input.
+ * @param {string}                     [props.placeholder]                Placeholder for the search control input
+ * @param {string}                     [props.className]                  The class name for this component
+ * @param {boolean}                    [props.disabled]                   Disables the component
+ * @param {Option[]}                   [props.options]                    Options to show in the component
+ * @param {string[]}                   [props.value]                      Selected values
+ * @param {number}                     [props.maxVisibleTags]             The maximum number of tags to show. Undefined, 0 or less than 0 evaluates to "Show All".
+ * @param {Function}                   [props.onChange]                   Callback when the selector changes
+ * @param {(visible: boolean) => void} [props.onDropdownVisibilityChange] Callback when the visibility of the dropdown options is changed.
+ * @return {JSX.Element} The component
+ */
+const TreeSelectControl = ( {
+	id,
+	label,
+	selectAllLabel = __( 'All', 'google-listings-and-ads' ),
+	help,
+	placeholder,
+	className,
+	disabled,
+	options = [],
+	value = [],
+	maxVisibleTags,
+	onChange = () => {},
+	onDropdownVisibilityChange = noop,
+} ) => {
+	let instanceId = useInstanceId( TreeSelectControl );
+	instanceId = id ?? instanceId;
+
+	const [ treeVisible, setTreeVisible ] = useState( false );
+	const [ nodesExpanded, setNodesExpanded ] = useState( [] );
+	const [ inputControlValue, setInputControlValue ] = useState( '' );
+
+	const controlRef = useRef();
+	const dropdownRef = useRef();
+	const onDropdownVisibilityChangeRef = useRef();
+	onDropdownVisibilityChangeRef.current = onDropdownVisibilityChange;
+
+	// We will save in a REF previous search filter queries to avoid re-query the tree and save performance
+	const cacheRef = useRef( { filteredOptionsMap: new Map() } );
+	cacheRef.current.expandedValues = nodesExpanded;
+	cacheRef.current.selectedValues = value;
+
+	const showTree = ! disabled && treeVisible;
+
+	const root =
+		selectAllLabel !== false
+			? {
+					label: selectAllLabel,
+					value: ROOT_VALUE,
+					children: options,
+			  }
+			: null;
+
+	const treeOptions = useIsEqualRefValue( root ? [ root ] : options );
+
+	const focusOutside = useFocusOutside( () => {
+		setTreeVisible( false );
+	} );
+
+	const filterQuery = inputControlValue.trim().toLowerCase();
+	const filter = filterQuery.length >= 3 ? filterQuery : '';
```

Comment:
Why do we use `filterQuery.length >= 3` here? Could we add a comment for it?

Reply:
Added here https://github.com/woocommerce/woocommerce/pull/33384/commits/d2ba89dac311ad645b5cde9ced56079ade72f444